chalk-solve/src/goal_builder.rs

```diff
@@ -0,0 +1,149 @@
+        &mut self,
+        clauses: CS,
+        goal: impl FnOnce(&mut Self) -> G,
+    ) -> Goal<I>
+    where
+        CS: IntoIterator<Item = C>,
+        C: CastTo<ProgramClause<I>>,
+        G: CastTo<Goal<I>>,
+    {
+        GoalData::Implies(
+            clauses.into_iter().casted(self.interner()).collect(),
+            goal(self).cast(self.interner()),
+        )
+        .intern(self.interner())
+    }
+
+    /// Given a bound value `binders` like `<P0..Pn> V`,
+    /// creates a goal `forall<Q0..Qn> { G }` where
+    /// the goal `G` is created by invoking a helper
+    /// function `body`.
+    ///
+    /// # Parameters to `body`
+    ///
+    /// `body` will be invoked with:
+    ///
+    /// * the goal builder `self`
+    /// * the substitution `Q0..Qn`
+    /// * the bound value `[P0..Pn => Q0..Qn] V` instantiated
+    ///   with the substitution
+    /// * the value `passthru`, appropriately shifted so that
+    ///   any debruijn indices within account for the new binder
+    ///
+    /// # Why is `body` a function and not a closure?
+    ///
+    /// This is to ensure that `body` doesn't accidentally reference
+    /// values from the environment whose debruijn indices do not
+    /// account for the new binder being created.
+    pub(crate) fn forall<G, B, P>(
+        &mut self,
+        binders: &Binders<B>,
+        passthru: P,
+        body: fn(&mut Self, Substitution<I>, &B, P::Result) -> G,
+    ) -> Goal<I>
+    where
+        B: Fold<I> + HasInterner<Interner = I>,
+        P: Fold<I>,
+        B::Result: std::fmt::Debug,
+        G: CastTo<Goal<I>>,
+    {
+        self.quantified(QuantifierKind::ForAll, binders, passthru, body)
+    }
+
+    /// Like [`GoalBuilder::forall`], but for a `exists<Q0..Qn> { G }` goal.
+    pub(crate) fn exists<G, B, P>(
+        &mut self,
+        binders: &Binders<B>,
+        passthru: P,
+        body: fn(&mut Self, Substitution<I>, &B, P::Result) -> G,
+    ) -> Goal<I>
+    where
+        B: Fold<I> + HasInterner<Interner = I>,
+        P: Fold<I>,
+        B::Result: std::fmt::Debug,
+        G: CastTo<Goal<I>>,
+    {
+        self.quantified(QuantifierKind::Exists, binders, passthru, body)
+    }
+
+    /// A combined helper functon for the various methods
+    /// to create `forall` and `exists` goals. See:
+    ///
+    /// * [`GoalBuilder::forall`]
+    /// * [`GoalBuilder::exists`]
+    ///
+    /// for details.
+    pub(crate) fn quantified<G, B, P>(
+        &mut self,
+        quantifier_kind: QuantifierKind,
+        binders: &Binders<B>,
+        passthru: P,
+        body: fn(&mut Self, Substitution<I>, &B, P::Result) -> G,
+    ) -> Goal<I>
+    where
+        B: Fold<I> + HasInterner<Interner = I>,
+        P: Fold<I>,
+        B::Result: std::fmt::Debug,
+        G: CastTo<Goal<I>>,
+    {
+        let interner = self.interner();
+        let bound_goal = binders.map_ref(|bound_value| {
+            let substitution: Substitution<I> = Substitution::from(
```

Comment:
I can move it out of the closure, but I'm not sure what you mean by "passing a `&Substitution` or clone" -- you'd rather I not give ownership of the `Substitution`? Or would you rather I clone the contents of the binder?

In some variant, I had the signature of the `body` be:

* `Substitution<I>`
* `B::Result` (i.e., contents of binder, instantiated with `Substitution`)
* `P::Result` (i.e., contents of passthru, shifted)

I changed to `&B` because, in fact, the substitution is always an identity substitution (in the earlier version, this wasn't always the case), so doing a substitution wasn't really necessary. But I think the above signature is kind of the most "obvious" signature, and I'd be reasonably happy to adopt it (since I think readability trumps most concerns here).